pkg/genericapiserver/openapi/openapi.go

```diff
@@ -0,0 +1,534 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+// Note: Any reference to swagger in this document is to swagger 1.2 spec.
+
+import (
+	"fmt"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/emicklei/go-restful"
+	"github.com/emicklei/go-restful/swagger"
+	"github.com/go-openapi/loads"
+	"github.com/go-openapi/spec"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/validate"
+	"k8s.io/kubernetes/pkg/util/json"
+)
+
+const (
+	// By convention, the Swagger specification file is named swagger.json
+	OpenAPIServePath = "/swagger.json"
+	OpenAPIVersion   = "2.0"
+)
+
+// Config is set of configuration for openAPI spec generation.
+type Config struct {
+	// SwaggerConfig is set of configuration for go-restful swagger spec generation. Currently
+	// openAPI implementation depends on go-restful to generate models.
+	SwaggerConfig *swagger.Config
+	// Info is general information about the API.
+	Info *spec.Info
+	// DefaultResponse will be used if an operation does not have any responses listed. It
+	// will show up as ... "responses" : {"default" : $DefaultResponse} in swagger spec.
+	DefaultResponse *spec.Response
+	// List of webservice's path prefixes to ignore
+	IgnorePrefixes []string
+}
+
+type openAPI struct {
+	config  *Config
+	swagger *spec.Swagger
+}
+
+// RegisterOpenAPIService registers a handler to provides standard OpenAPI specification.
+func RegisterOpenAPIService(config *Config, containers *restful.Container) (err error) {
+	var _ = loads.Spec
+	var _ = strfmt.ParseDuration
+	var _ = validate.FormatOf
+	o := openAPI{
+		config: config,
+	}
+	err = o.buildSwaggerSpec()
+	if err != nil {
+		return err
+	}
+	containers.ServeMux.HandleFunc(OpenAPIServePath, func(w http.ResponseWriter, r *http.Request) {
+		resp := restful.NewResponse(w)
+		if r.URL.Path != OpenAPIServePath {
+			resp.WriteErrorString(http.StatusNotFound, "Path not found!")
+		}
+		resp.WriteAsJson(o.swagger)
+	})
+	return nil
+}
+
+func (o *openAPI) buildSwaggerSpec() (err error) {
+	if o.swagger != nil {
+		return fmt.Errorf("Swagger spec is already built. Duplicate call to buildSwaggerSpec is not allowed.")
+	}
+	definitions, err := o.buildModels()
+	if err != nil {
+		return err
+	}
+	paths, err := o.buildPaths()
+	if err != nil {
+		return err
+	}
+	o.swagger = &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Swagger:     OpenAPIVersion,
+			Definitions: definitions,
+			Paths:       &paths,
+			Info:        o.config.Info,
+		},
+	}
+	return nil
+}
+
+func (o *openAPI) buildModels() (definitions spec.Definitions, err error) {
```

Comment:
No entirely sure what you mean by that. I will add a godoc comment for these methods, though they are private and (I think somehow) self explanatory by their signature.